File: HWAppFoundation/src/main/java/com/hw/hwdroid/foundation/utils/toast/ToastUtils.java
package com.hw.hwdroid.foundation.utils.toast;

import android.content.Context;
import android.support.annotation.IntDef;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import com.hw.hwdroid.foundation.R;
import com.hw.hwdroid.foundation.utils.StringUtils;
import com.orhanobut.logger.Logger;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Consumer;

/**
 * 自定义Toast Util
 *
 * @author chenjian
 * @date 2014-1-15
 */

public class ToastUtils {

    private CustomToast customToast;

    private static ToastUtils toastUtil;

    private ToastUtils(Context _Context) {
        customToast = new CustomToast(_Context);
    }

    public synchronized static ToastUtils instance(Context _Context) {
        synchronized (ToastUtils.class) {
            if (toastUtil == null) {
                toastUtil = new ToastUtils(_Context);
            }
        }

        return toastUtil;
    }

    public synchronized static void releaseToast() {
        toastUtil = null;
    }

    /**
     * 自定义短时间toast
     *
     * @param msgResId
     */
    public synchronized void showCustomToastShort(@StringRes final int msgResId) {
        customToast.showShort(msgResId);
    }

    /**
     * 自定义短时间toast
     *
     * @param text
     */
    public synchronized void showCustomToastShort(@NonNull final String text) {
        customToast.showShort(text);
    }

    /**
     * 自定义长时间toast
     *
     * @param msgResId
     */
    public synchronized void showCustomToastLong(@StringRes final int msgResId) {
        customToast.showLong(msgResId);
    }

    /**
     * 自定义长时间toast
     *
     * @param text
     */
    public synchronized void showCustomToastLong(@NonNull final String text) {
        customToast.showLong(text);
    }

    /**
     * 自定义toast
     *
     * @param msgResId
     * @param duration
     */
    public synchronized void showCustomToast(@StringRes final int msgResId, final int duration) {
        customToast.show(msgResId, duration);
    }

    public synchronized void showCustomToast(@NonNull final String msg, final int duration) {
        customToast.show(msg, duration);
    }

    /**
     * 自定义toas
     *
     * @param msg
     * @param duration
     * @param judgeAppIsForeground 判断APP是否在前台
     */
    public synchronized void showCustomToast(@NonNull final String msg, final int duration, final boolean judgeAppIsForeground) {
        customToast.show(msg, duration, judgeAppIsForeground);
    }

    /**
     * 显示冒泡提示
     *
     * @param context
     * @param msg
     */
    public static void show(@NonNull Context context, @NonNull String msg) {
        show(context, msg, Gravity.CENTER, 0, 0, Toast.LENGTH_SHORT);
    }

    /**
     * 显示冒泡提示
     *
     * @param context
     * @param msg
     * @param action  延迟执行的动作
     */
    public static void show(@NonNull Context context, @NonNull String msg, Consumer<Integer> action) {
        show(context, msg, Gravity.CENTER, 0, 0, Toast.LENGTH_SHORT, action);
    }

    /**
     * 显示冒泡提示
     *
     * @param context
     * @param msg
     * @param defMsgResId msg为空时使用defMsgResId
     */
    public static void show2(@NonNull Context context, @NonNull String msg, @StringRes final int defMsgResId) {
        if (StringUtils.isNullOrWhiteSpace(msg)) {
            show(context, defMsgResId);
        } else {
            show(context, msg, Gravity.CENTER, 0, 0, Toast.LENGTH_SHORT);
        }
    }

    /**
     * 显示冒泡提示
     *
     * @param context
     * @param msgResId
     * @param action   延迟执行的动作
     */
    public static void show(@NonNull Context context, @StringRes final int msgResId, Consumer<Integer> action) {
        String msg = context.getString(msgResId);
        show(context, msg, Gravity.CENTER, 0, 0, Toast.LENGTH_SHORT, action);
    }

    /**
     * 显示冒泡提示
     *
     * @param context
     * @param msgResId
     */
    public static void show(@NonNull Context context, @StringRes final int msgResId) {
        show(context, msgResId, null);
    }

    /**
     * 显示冒泡提示
     *
     * @param context
     * @param msg
     */
    public static void showLong(@NonNull Context context, String msg) {
        show(context, msg, Gravity.CENTER, 0, 0, Toast.LENGTH_LONG);
    }

    /**
     * 显示冒泡提示
     *
     * @param context
     * @param msgResId
     */
    public static void showLong(@NonNull Context context, @StringRes final int msgResId) {
        String msg = context.getString(msgResId);
        show(context, msg, Gravity.CENTER, 0, 0, Toast.LENGTH_LONG);
    }

    /**
     * 显示冒泡提示
     *
     * @param context
     * @param msgResId
     * @param duration
     */
    public static void show(@NonNull Context context, @StringRes final int msgResId, int duration) {
        show(context, msgResId, duration, null);
    }

    /**
     * 显示冒泡提示
     *
     * @param context
     * @param msgResId
     * @param duration
     * @param action   延迟执行的动作
     */
    public static void show(@NonNull Context context, @StringRes final int msgResId, int duration, Consumer<Integer> action) {
        String msg = context.getString(msgResId);
        show(context, msg, Gravity.CENTER, 0, 0, duration, action);
    }

    /**
     * 显示冒泡提示
     *
     * @param context
     * @param msg
     * @param duration
     */
    public static void show(@NonNull Context context, @NonNull String msg, int duration) {
        show(context, msg, Gravity.CENTER, 0, 0, duration);
    }

    /**
     * 显示冒泡提示
     *
     * @param context
     * @param msg
     * @param gravity
     * @param duration
     */
    public static void show(@NonNull Context context, @NonNull String msg, int gravity, int duration) {
        show(context, msg, gravity, 0, 0, duration);
    }

    /**
     * 显示冒泡提示
     *
     * @param context
     * @param msg
     * @param gravity  Gravity.CENTER
     * @param xOffset
     * @param yOffset
     * @param duration
     */
    public static void show(@NonNull Context context, @NonNull String msg, int gravity, int xOffset, int yOffset, int duration) {
        show(context, msg, gravity, xOffset, yOffset, duration, null);
    }

    @IntDef({Toast.LENGTH_SHORT, Toast.LENGTH_LONG})
    @Retention(RetentionPolicy.SOURCE)
    public @interface Duration {
    }

    /**
     * 显示冒泡提示
     *
     * @param context
     * @param msg
     * @param gravity  Gravity.CENTER
     * @param xOffset
     * @param yOffset
     * @param duration
     * @param action   延迟执行的动作
     */
    public static void show(@NonNull Context context,
                            @NonNull String msg, int gravity, int xOffset, int yOffset, @Duration int duration, Consumer<Integer> action) {
        if (null == context || StringUtils.isNullOrWhiteSpace(msg)) {
            return;
        }

        Toast toast = new Toast(context);
        View view = LayoutInflater.from(context.getApplicationContext()).inflate(R.layout.ab__toast_layout, null);
        ((TextView) view.findViewById(R.id.my_toast)).setText(msg);
        toast.setView(view);

        toast.setGravity(gravity, xOffset, yOffset);

        toast.setDuration(duration);
        toast.show();

        if (null != action) {
            Observable.just(1).delay(duration + 1, TimeUnit.SECONDS).observeOn(AndroidSchedulers.mainThread()).subscribe(action, onError -> Logger.e(onError));
        }
    }

}

File: HWAppFoundation/src/main/java/com/hw/hwdroid/foundation/utils/toast/ToastUtils.kt
package com.hw.hwdroid.foundation.utils.toast

import android.content.Context
import android.support.annotation.IntDef
import android.support.annotation.StringRes
import android.view.Gravity
import android.view.LayoutInflater
import android.widget.TextView
import android.widget.Toast
import com.hw.hwdroid.foundation.R
import com.hw.hwdroid.foundation.app.HWFoundationContext
import com.hw.hwdroid.foundation.utils.ResourceUtils
import com.hw.hwdroid.foundation.utils.StringUtils
import com.orhanobut.logger.Logger
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.Consumer
import java.lang.annotation.Retention
import java.lang.annotation.RetentionPolicy
import java.util.concurrent.TimeUnit

/**
 * 自定义Toast Util
 *
 * @author chenJ
 * @date 2014-1-15
 */

class ToastUtils private constructor(_Context: Context) {

    private val customToast: CustomToast = CustomToast(_Context)

    /**
     * 自定义toast
     *
     * @param messageResId
     * @param duration
     * @param judgeAppIsForeground  判断APP是否在前台
     * @param action                延迟执行的动作
     */
    @Synchronized fun showCustomToast(@StringRes messageResId: Int, @ToastUtils.Duration duration: Int = Toast.LENGTH_SHORT, judgeAppIsForeground: Boolean = true, gravity: Int = Gravity.CENTER, xOffset: Int = 0, yOffset: Int = 0, action: Consumer<Int>?) {
        customToast.show(messageResId, duration, judgeAppIsForeground, gravity, xOffset, yOffset, action)
    }

    /**
     * 自定义toast
     *
     * @param message
     * @param duration
     * @param judgeAppIsForeground  判断APP是否在前台
     * @param action                延迟执行的动作
     */
    @Synchronized fun showCustomToast(message: CharSequence, @ToastUtils.Duration duration: Int = Toast.LENGTH_SHORT, judgeAppIsForeground: Boolean = true, gravity: Int = Gravity.CENTER, xOffset: Int = 0, yOffset: Int = 0, action: Consumer<Int>?) {
        customToast.show(message, duration, judgeAppIsForeground, gravity, xOffset, yOffset, action)
    }

    @IntDef(Toast.LENGTH_SHORT.toLong(), Toast.LENGTH_LONG.toLong())
    @Retention(RetentionPolicy.SOURCE)
    annotation class Duration

    companion object {

        private var toastUtil: ToastUtils? = null

        @Synchronized fun instance(): ToastUtils? {
            synchronized(ToastUtils::class.java) {
                if (toastUtil == null) {
                    val context = HWFoundationContext.getApplicationContext()
                    if (context != null) {
                        toastUtil = ToastUtils(context)
                    }
                }
            }

            return toastUtil
        }

        @Synchronized fun releaseToast() {
            toastUtil = null
        }

        /**
         *
         */
        @JvmOverloads fun show(context: Context, @StringRes messageResId: Int, @Duration duration: Int = Toast.LENGTH_SHORT, gravity: Int = Gravity.CENTER, xOffset: Int = 0, yOffset: Int = 0, action: Consumer<Int>? = null) {
            show(context, ResourceUtils.getString(context, messageResId, ""), duration, gravity, xOffset, yOffset, action)
        }

        /**
         * 显示冒泡提示
         *
         * @param context
         * @param message
         * @param gravity       Gravity.CENTER
         * @param xOffset       0
         * @param yOffset       0
         * @param duration           duration
         * @param action        延迟执行的动作
         */
        @JvmOverloads fun show(context: Context, message: CharSequence, @Duration duration: Int = Toast.LENGTH_SHORT, gravity: Int = Gravity.CENTER, xOffset: Int = 0, yOffset: Int = 0, action: Consumer<Int>? = null) {
            if (StringUtils.isNullOrWhiteSpace(message)) {
                return
            }

            val toast = Toast(context)
            val view = LayoutInflater.from(context.applicationContext).inflate(R.layout.ab__toast_layout, null)
            (view.findViewById(R.id.my_toast) as? TextView)?.text = message
            toast.view = view

            toast.setGravity(gravity, xOffset, yOffset)
            toast.duration = duration
            toast.show()

            if (action != null) {
                Observable.just(1).delay((duration + 1).toLong(), TimeUnit.SECONDS).observeOn(AndroidSchedulers.mainThread()).subscribe(action, Consumer { onError -> Logger.e(onError) })
            }
        }
    }

}